fedimg/utils.py

```diff
@@ -88,16 +89,93 @@ def ssh_connection_works(username, ip, keypath):
         works = True
     except (paramiko.BadHostKeyException,
             paramiko.AuthenticationException,
-            paramiko.SSHException, socket.error) as e:
+            paramiko.SSHException, socket.error):
         pass
     ssh.close()
     return works
 
 
-def safeget(dct, *keys):
+def get_value_from_dict(_dict, *keys):
     for key in keys:
         try:
-            dct = dct[key]
+            _dict = _dict[key]
         except KeyError:
             return None
-    return dct
+    return _dict
+
+
+def external_run_command(command):
+    ret = subprocess.Popen(' '.join(command), shell=True, stdin=subprocess.PIPE,
+                           stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                           close_fds=True)
+    out, err = ret.communicate()
+    retcode = ret.returncode
+    return out, err, retcode
+
+
+def get_item_from_regex(output, regex):
+    match = re.search(regex, output)
+    if match is None:
+        return ''
+    else:
+        return match.group(1)
+
+
+def get_file_name_image(image_url):
+    return image_url.split('/')[-1]
+
+
+def get_source_from_image(image_url):
+    tmpdir = tempfile.mkdtemp()
+    file_name = get_file_name_image(image_url)
+    file_path = os.path.join(tmpdir, file_name)
+
+    log.info("[PREP] Preparing temporary directory for download: %r" % tmpdir)
+    output, error, retcode = external_run_command([
+        'wget',
+        image_url,
+        '-P',
+        tmpdir
+    ])
+
+    return file_path
+
+
+def get_volume_type_from_image(image, region):
+    return image.extra['block_device_mapping'][0]['ebs']['volume_type']
+
+
+def get_virt_type_from_image(image):
+    device_name = image.extra['block_device_mapping'][0]['device_name']
+
+    if device_name.endswith('sda1'):
+        return 'hvm'
+    else:
+        return 'paravirtual'
+
+
+def get_image_name_from_image(image_url, virt_type='', region='', respin='0',
+                              volume_type=''):
+
+    file_name = get_file_name_image(image_url)
+    build_name = file_name.replace('.raw.xz', '')
+
+    return '-'.join(filter(lambda x: x, [
+        build_name,
+        virt_type,
+        region,
+        volume_type,
+        respin]
+    ))
```

Comment:
I'm not sure what this does, but I feel like it could be simplified, no?

Reply:
This generates the image name, and appends the properties like `region`, `volume_type` etc to the image. If not present then it just skips that instead of joining an empty string. The sole purpose of the lambda there is to check if the string is empty of not.

Reply:
Why not doing: 
````
return '-'.join([x for x in [build_name, virt_type, region, volume_type, respin] if x ])
````
It's more common and imho easier to read :)